File: wg-wss/client-go/ws.go
// Client side of RFC 6455, standard library only, so the binary
// cross-compiles anywhere Go runs without fetching a single module.
package main

import (
	"bufio"
	"crypto/rand"
	"crypto/sha1"
	"encoding/base64"
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"strings"
	"sync"
)

const (
	opCont  = 0x0
	opText  = 0x1
	opBin   = 0x2
	opClose = 0x8
	opPing  = 0x9
	opPong  = 0xA

	wsGUID     = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
	maxMessage = 128 * 1024
)

// wsConn is a WebSocket connection with independent read and write locks, so
// a keepalive ping never interleaves with a packet being written.
type wsConn struct {
	conn net.Conn
	br   *bufio.Reader
	mu   sync.Mutex
}

func acceptKey(key string) string {
	sum := sha1.Sum([]byte(key + wsGUID))
	return base64.StdEncoding.EncodeToString(sum[:])
}

// wsDial performs the opening handshake over an already-connected net.Conn.
func wsDial(conn net.Conn, hostHeader, path string, headers map[string]string) (*wsConn, error) {
	nonce := make([]byte, 16)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	key := base64.StdEncoding.EncodeToString(nonce)

	var req strings.Builder
	fmt.Fprintf(&req, "GET %s HTTP/1.1\r\n", path)
	fmt.Fprintf(&req, "Host: %s\r\n", hostHeader)
	req.WriteString("Upgrade: websocket\r\n")
	req.WriteString("Connection: Upgrade\r\n")
	fmt.Fprintf(&req, "Sec-WebSocket-Key: %s\r\n", key)
	req.WriteString("Sec-WebSocket-Version: 13\r\n")
	req.WriteString("User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 " +
		"(KHTML, like Gecko) Chrome/125.0 Safari/537.36\r\n")
	for name, value := range headers {
		fmt.Fprintf(&req, "%s: %s\r\n", name, value)
	}
	req.WriteString("\r\n")
	if _, err := io.WriteString(conn, req.String()); err != nil {
		return nil, err
	}

	br := bufio.NewReaderSize(conn, 4096)
	status, err := br.ReadString('\n')
	if err != nil {
		return nil, err
	}
	if !strings.Contains(status, " 101") {
		return nil, fmt.Errorf("handshake rejected: %s", strings.TrimSpace(status))
	}
	var gotAccept string
	for {
		line, err := br.ReadString('\n')
		if err != nil {
			return nil, err
		}
		line = strings.TrimRight(line, "\r\n")
		if line == "" {
			break
		}
		name, value, _ := strings.Cut(line, ":")
		if strings.EqualFold(strings.TrimSpace(name), "Sec-WebSocket-Accept") {
			gotAccept = strings.TrimSpace(value)
		}
	}
	if gotAccept != acceptKey(key) {
		return nil, fmt.Errorf("bad Sec-WebSocket-Accept")
	}
	return &wsConn{conn: conn, br: br}, nil
}

func (c *wsConn) writeFrame(opcode byte, payload []byte) error {
	n := len(payload)
	head := make([]byte, 0, 14)
	head = append(head, 0x80|opcode)
	switch {
	case n < 126:
		head = append(head, 0x80|byte(n))
	case n < 65536:
		head = append(head, 0x80|126, byte(n>>8), byte(n))
	default:
		head = append(head, 0x80|127)
		var ext [8]byte
		binary.BigEndian.PutUint64(ext[:], uint64(n))
		head = append(head, ext[:]...)
	}
	var mask [4]byte
	if _, err := rand.Read(mask[:]); err != nil {
		return err
	}
	head = append(head, mask[:]...)

	// Clients must mask; do it into a scratch buffer, never in place.
	body := make([]byte, n)
	for i := 0; i < n; i++ {
		body[i] = payload[i] ^ mask[i&3]
	}

	c.mu.Lock()
	defer c.mu.Unlock()
	if _, err := c.conn.Write(append(head, body...)); err != nil {
		return err
	}
	return nil
}

func (c *wsConn) Send(data []byte) error { return c.writeFrame(opBin, data) }
func (c *wsConn) Ping() error            { return c.writeFrame(opPing, nil) }

// Recv returns the next binary message, answering control frames itself.
func (c *wsConn) Recv() ([]byte, error) {
	var buf []byte
	for {
		var head [2]byte
		if _, err := io.ReadFull(c.br, head[:]); err != nil {
			return nil, err
		}
		if head[0]&0x70 != 0 {
			return nil, fmt.Errorf("reserved bits set")
		}
		fin := head[0]&0x80 != 0
		opcode := head[0] & 0x0F
		masked := head[1]&0x80 != 0
		n := int(head[1] & 0x7F)
		switch n {
		case 126:
			var ext [2]byte
			if _, err := io.ReadFull(c.br, ext[:]); err != nil {
				return nil, err
			}
			n = int(binary.BigEndian.Uint16(ext[:]))
		case 127:
			var ext [8]byte
			if _, err := io.ReadFull(c.br, ext[:]); err != nil {
				return nil, err
			}
			size := binary.BigEndian.Uint64(ext[:])
			if size > maxMessage {
				return nil, fmt.Errorf("message of %d bytes exceeds limit", size)
			}
			n = int(size)
		}
		if n > maxMessage {
			return nil, fmt.Errorf("message of %d bytes exceeds limit", n)
		}
		var mask [4]byte
		if masked {
			if _, err := io.ReadFull(c.br, mask[:]); err != nil {
				return nil, err
			}
		}
		payload := make([]byte, n)
		if _, err := io.ReadFull(c.br, payload); err != nil {
			return nil, err
		}
		if masked {
			for i := range payload {
				payload[i] ^= mask[i&3]
			}
		}

		switch opcode {
		case opClose:
			return nil, io.EOF
		case opPing:
			if err := c.writeFrame(opPong, payload); err != nil {
				return nil, err
			}
			continue
		case opPong:
			continue
		case opBin, opText, opCont:
			buf = append(buf, payload...)
			if len(buf) > maxMessage {
				return nil, fmt.Errorf("fragmented message exceeds limit")
			}
			if fin {
				return buf, nil
			}
		default:
			return nil, fmt.Errorf("unknown opcode 0x%x", opcode)
		}
	}
}

func (c *wsConn) Close() error {
	_ = c.writeFrame(opClose, []byte{0x03, 0xE8}) // 1000, normal closure
	return c.conn.Close()
}

File: wg-wss/wgws/ws.py
"""Minimal RFC 6455 WebSocket implementation on top of asyncio streams.

Only what the tunnel needs: binary messages, ping/pong keepalive, close
handshake, client-side masking. No third-party dependencies.
"""

import asyncio
import base64
import hashlib
import os
import struct

GUID = b"258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

OP_CONT = 0x0
OP_TEXT = 0x1
OP_BIN = 0x2
OP_CLOSE = 0x8
OP_PING = 0x9
OP_PONG = 0xA

CONTROL_OPS = (OP_CLOSE, OP_PING, OP_PONG)

# A WireGuard datagram never exceeds the interface MTU by much; anything
# larger than this is either a bug or someone poking at the port.
MAX_MESSAGE = 128 * 1024


class WSError(Exception):
    """Protocol violation."""


class WSClosed(Exception):
    """Peer closed the connection (cleanly or not)."""


def accept_key(key):
    """Sec-WebSocket-Accept value for a given Sec-WebSocket-Key."""
    digest = hashlib.sha1(key.encode("ascii", "strict") + GUID).digest()
    return base64.b64encode(digest).decode("ascii")


def apply_mask(data, mask):
    """XOR `data` with the 4-byte `mask` (same operation both directions)."""
    n = len(data)
    if n == 0:
        return data
    key = int.from_bytes(mask * ((n + 3) // 4), "big") >> (8 * ((-n) % 4))
    return (int.from_bytes(data, "big") ^ key).to_bytes(n, "big")


def build_frame(opcode, payload=b"", mask=False):
    """Serialize a single final frame."""
    b1 = 0x80 | opcode
    n = len(payload)
    flag = 0x80 if mask else 0x00
    if n < 126:
        head = struct.pack("!BB", b1, flag | n)
    elif n < 65536:
        head = struct.pack("!BBH", b1, flag | 126, n)
    else:
        head = struct.pack("!BBQ", b1, flag | 127, n)
    if not mask:
        return head + payload
    key = os.urandom(4)
    return head + key + apply_mask(payload, key)


class WSConnection:
    """A WebSocket connection over an established asyncio stream pair."""

    def __init__(self, reader, writer, mask_out, max_message=MAX_MESSAGE):
        self.reader = reader
        self.writer = writer
        self.mask_out = mask_out
        self.max_message = max_message
        self.closed = False
        self.bytes_in = 0
        self.bytes_out = 0
        self._send_lock = asyncio.Lock()

    # -- sending ---------------------------------------------------------
    async def _send_frame(self, opcode, payload=b""):
        if self.closed:
            raise WSClosed("connection is closed")
        frame = build_frame(opcode, payload, self.mask_out)
        async with self._send_lock:
            self.writer.write(frame)
            await self.writer.drain()
        self.bytes_out += len(payload)

    async def send(self, data):
        await self._send_frame(OP_BIN, data)

    async def ping(self, data=b""):
        await self._send_frame(OP_PING, data)

    # -- receiving -------------------------------------------------------
    async def _read_frame(self):
        head = await self.reader.readexactly(2)
        b1, b2 = head[0], head[1]
        if b1 & 0x70:
            raise WSError("reserved bits set")
        fin = bool(b1 & 0x80)
        opcode = b1 & 0x0F
        masked = bool(b2 & 0x80)
        n = b2 & 0x7F
        if n == 126:
            n = struct.unpack("!H", await self.reader.readexactly(2))[0]
        elif n == 127:
            n = struct.unpack("!Q", await self.reader.readexactly(8))[0]
        if opcode in CONTROL_OPS:
            if not fin:
                raise WSError("fragmented control frame")
            if n > 125:
                raise WSError("oversized control frame")
        if n > self.max_message:
            raise WSError("message of %d bytes exceeds limit" % n)
        key = await self.reader.readexactly(4) if masked else None
        payload = await self.reader.readexactly(n) if n else b""
        if key is not None:
            payload = apply_mask(payload, key)
        return fin, opcode, payload

    async def recv(self):
        """Return the next binary message; control frames are handled here."""
        buf = b""
        frag_op = None
        while True:
            try:
                fin, opcode, payload = await self._read_frame()
            except (asyncio.IncompleteReadError, ConnectionError) as exc:
                self.closed = True
                raise WSClosed(str(exc) or "stream ended")
            if opcode == OP_CLOSE:
                self.closed = True
                raise WSClosed("peer sent close")
            if opcode == OP_PING:
                await self._send_frame(OP_PONG, payload)
                continue
            if opcode == OP_PONG:
                continue
            if opcode == OP_CONT:
                if frag_op is None:
                    raise WSError("continuation without start frame")
            elif opcode in (OP_BIN, OP_TEXT):
                if frag_op is not None:
                    raise WSError("new message before previous one finished")
                frag_op = opcode
            else:
                raise WSError("unknown opcode 0x%x" % opcode)
            buf += payload
            if len(buf) > self.max_message:
                raise WSError("fragmented message exceeds limit")
            if fin:
                self.bytes_in += len(buf)
                return buf

    # -- teardown --------------------------------------------------------
    async def close(self, code=1000):
        if not self.closed:
            self.closed = True
            try:
                await asyncio.wait_for(
                    self._force_frame(OP_CLOSE, struct.pack("!H", code)), 2
                )
            except (OSError, asyncio.TimeoutError, ConnectionError):
                pass
        try:
            self.writer.close()
            await asyncio.wait_for(self.writer.wait_closed(), 2)
        except (OSError, asyncio.TimeoutError, ConnectionError):
            pass

    async def _force_frame(self, opcode, payload):
        frame = build_frame(opcode, payload, self.mask_out)
        async with self._send_lock:
            self.writer.write(frame)
            await self.writer.drain()


async def read_http_head(reader, limit=16384, timeout=15):
    """Read an HTTP request/response head and return (start_line, headers)."""
    raw = await asyncio.wait_for(reader.readuntil(b"\r\n\r\n"), timeout)
    if len(raw) > limit:
        raise WSError("HTTP head too large")
    lines = raw.decode("latin-1").split("\r\n")
    start = lines[0]
    headers = {}
    for line in lines[1:]:
        if not line:
            continue
        name, _, value = line.partition(":")
        headers[name.strip().lower()] = value.strip()
    return start, headers


async def client_handshake(reader, writer, host, path, extra_headers=None):
    """Perform the client side of the opening handshake."""
    key = base64.b64encode(os.urandom(16)).decode("ascii")
    lines = [
        "GET %s HTTP/1.1" % path,
        "Host: %s" % host,
        "Upgrade: websocket",
        "Connection: Upgrade",
        "Sec-WebSocket-Key: %s" % key,
        "Sec-WebSocket-Version: 13",
        "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/125.0 Safari/537.36",
    ]
    for name, value in (extra_headers or {}).items():
        lines.append("%s: %s" % (name, value))
    writer.write(("\r\n".join(lines) + "\r\n\r\n").encode("latin-1"))
    await writer.drain()

    start, headers = await read_http_head(reader)
    parts = start.split(" ", 2)
    if len(parts) < 2 or parts[1] != "101":
        raise WSError("handshake rejected: %s" % start)
    if headers.get("upgrade", "").lower() != "websocket":
        raise WSError("missing Upgrade header in response")
    if headers.get("sec-websocket-accept", "") != accept_key(key):
        raise WSError("bad Sec-WebSocket-Accept")
    return WSConnection(reader, writer, mask_out=True)


def is_websocket_upgrade(headers):
    return (
        headers.get("upgrade", "").lower() == "websocket"
        and "upgrade" in headers.get("connection", "").lower()
        and bool(headers.get("sec-websocket-key"))
    )


def server_accept_response(headers):
    """101 response bytes for a validated upgrade request."""
    return (
        "HTTP/1.1 101 Switching Protocols\r\n"
        "Upgrade: websocket\r\n"
        "Connection: Upgrade\r\n"
        "Sec-WebSocket-Accept: %s\r\n\r\n" % accept_key(headers["sec-websocket-key"])
    ).encode("latin-1")
